File: apps/visualizer/backend/api/schemas/providers.py
"""Providers API response models — single source of truth for provider shapes."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, RootModel

ProviderModelSource = Literal['config', 'discovered', 'user']


class Provider(BaseModel):
    model_config = ConfigDict(extra='forbid')

    id: str
    name: str
    available: bool
    tool_calling: bool


class ProviderModel(BaseModel):
    model_config = ConfigDict(extra='forbid')

    id: str
    name: str
    source: ProviderModelSource
    vision: bool | None = None


class ProviderDefaultsEntry(BaseModel):
    model_config = ConfigDict(extra='forbid')

    provider: str
    model: str | None = None


class ProviderDefaults(BaseModel):
    model_config = ConfigDict(extra='forbid')

    vision_comparison: ProviderDefaultsEntry
    description: ProviderDefaultsEntry


class DescriptionModel(BaseModel):
    model_config = ConfigDict(extra='forbid')

    provider_id: str
    provider_name: str
    model_id: str
    model_name: str
    tool_calling: bool


class DescriptionModelsResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    models: list[DescriptionModel]
    default_provider: str | None = None
    default_model: str | None = None


class FallbackOrderResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    order: list[str]


class ProviderHealthResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    reachable: bool
    error: str | None = None


class ProviderDeletedResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    deleted: bool


class ProviderReorderSuccessResponse(BaseModel):
    model_config = ConfigDict(extra='forbid')

    success: bool


class ProviderModelsListResponse(RootModel[list[ProviderModel]]):
    """``GET /api/providers/<id>/models`` response body."""


class ProviderListResponse(RootModel[list[Provider]]):
    """``GET /api/providers/`` response body."""

File: apps/visualizer/backend/src/api/schemas/providers.ts
/** Providers API models. */
import { z } from '@hono/zod-openapi';

export const ProviderModelSource = z.enum(['config', 'discovered', 'user']);

export const Provider = z
  .object({
    id: z.string(),
    name: z.string(),
    available: z.boolean(),
    tool_calling: z.boolean(),
  })
  .strict()
  .openapi('Provider');

export const ProviderModel = z
  .object({
    id: z.string(),
    name: z.string(),
    source: ProviderModelSource,
    vision: z.boolean().nullish(),
  })
  .strict()
  .openapi('ProviderModel');

export const ProviderDefaultsEntry = z
  .object({ provider: z.string(), model: z.string().nullish() })
  .strict()
  .openapi('ProviderDefaultsEntry');

export const ProviderDefaults = z
  .object({ description: ProviderDefaultsEntry })
  .strict()
  .openapi('ProviderDefaults');

export const DescriptionModel = z
  .object({
    provider_id: z.string(),
    provider_name: z.string(),
    model_id: z.string(),
    model_name: z.string(),
    tool_calling: z.boolean(),
  })
  .strict()
  .openapi('DescriptionModel');

export const DescriptionModelsResponse = z
  .object({
    models: z.array(DescriptionModel),
    default_provider: z.string().nullish(),
    default_model: z.string().nullish(),
  })
  .strict()
  .openapi('DescriptionModelsResponse');

export const FallbackOrderResponse = z
  .object({ order: z.array(z.string()) })
  .strict()
  .openapi('FallbackOrderResponse');

export const ProviderHealthResponse = z
  .object({ reachable: z.boolean(), error: z.string().nullish() })
  .strict()
  .openapi('ProviderHealthResponse');

export const ProviderDeletedResponse = z
  .object({ deleted: z.boolean() })
  .strict()
  .openapi('ProviderDeletedResponse');

export const ProviderReorderSuccessResponse = z
  .object({ success: z.boolean() })
  .strict()
  .openapi('ProviderReorderSuccessResponse');

/**
 * Top-level JSON arrays, not wrapped objects — matching the wire format of
 * `GET /api/providers/` and `GET /api/providers/{id}/models`.
 */
export const ProviderListResponse = z.array(Provider).openapi('ProviderListResponse');
export const ProviderModelsListResponse = z
  .array(ProviderModel)
  .openapi('ProviderModelsListResponse');

export type ProviderModel = z.infer<typeof ProviderModel>;
